File: packages/inputlayer-py/src/inputlayer/_ast.py
"""Internal AST nodes for expression trees compiled to Datalog."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any


# ── Base ──────────────────────────────────────────────────────────────

@dataclass(frozen=True)
class Expr:
    """Base class for all expression AST nodes."""


@dataclass(frozen=True)
class BoolExpr:
    """Base class for boolean expression AST nodes (conditions)."""


# ── Leaf nodes ────────────────────────────────────────────────────────

@dataclass(frozen=True)
class Column(Expr):
    """Reference to a relation column."""
    relation: str
    name: str
    ref_alias: str | None = None  # For self-join disambiguation

    @property
    def qualified(self) -> str:
        prefix = self.ref_alias or self.relation
        return f"{prefix}.{self.name}"


@dataclass(frozen=True)
class Literal(Expr):
    """A constant value."""
    value: Any  # int, float, str, bool, list (vector), None


# ── Arithmetic ────────────────────────────────────────────────────────

@dataclass(frozen=True)
class Arithmetic(Expr):
    """Binary arithmetic: +, -, *, /, %."""
    op: str  # "+", "-", "*", "/", "%"
    left: Expr
    right: Expr


# ── Function call ─────────────────────────────────────────────────────

@dataclass(frozen=True)
class FuncCall(Expr):
    """Built-in function call: distance(V1, V2), upper(S), etc."""
    name: str
    args: tuple[Expr, ...] = ()


# ── Aggregation ───────────────────────────────────────────────────────

@dataclass(frozen=True)
class AggExpr(Expr):
    """Aggregation expression: count<X>, sum<X>, top_k<k, ...>, etc."""
    func: str  # "count", "sum", "min", "max", "avg", "count_distinct",
    # "top_k", "top_k_threshold", "within_radius"
    column: Expr | None = None  # The aggregated column (None for count(*))
    params: tuple[Any, ...] = ()  # Extra params (k, threshold, etc.)
    passthrough: tuple[Expr, ...] = ()  # Passthrough columns
    order_column: Expr | None = None  # For top_k: the ordering column
    desc: bool = True  # For top_k: descending order


# ── Ordering ──────────────────────────────────────────────────────────

@dataclass(frozen=True)
class OrderedColumn(Expr):
    """A column with sort direction."""
    column: Expr
    descending: bool = False


# ── Boolean expressions ──────────────────────────────────────────────

@dataclass(frozen=True)
class Comparison(BoolExpr):
    """Binary comparison: ==, !=, <, <=, >, >=."""
    op: str  # "=", "!=", "<", "<=", ">", ">="
    left: Expr
    right: Expr


@dataclass(frozen=True)
class And(BoolExpr):
    """Logical AND of two conditions (Datalog comma)."""
    left: BoolExpr
    right: BoolExpr


@dataclass(frozen=True)
class Or(BoolExpr):
    """Logical OR - requires splitting into multiple queries."""
    left: BoolExpr
    right: BoolExpr


@dataclass(frozen=True)
class Not(BoolExpr):
    """Negation: !relation(X, Y) in Datalog."""
    operand: BoolExpr


@dataclass(frozen=True)
class InExpr(BoolExpr):
    """Membership test: Column appears in another relation."""
    column: Expr
    target_column: Expr


@dataclass(frozen=True)
class NegatedIn(BoolExpr):
    """Negated membership test."""
    column: Expr
    target_column: Expr


@dataclass(frozen=True)
class MatchExpr(BoolExpr):
    """Multi-column negation/existence check against a relation."""
    relation: str
    bindings: dict[str, Expr]  # target_col -> source expr
    negated: bool = False

File: packages/inputlayer-js/src/ast.ts
/**
 * Internal AST nodes for expression trees compiled to IQL.
 */

// ── Base types ──────────────────────────────────────────────────────

/** Base interface for all expression AST nodes. */
export interface Expr {
  readonly _tag: string;
}

/** Base interface for boolean expression AST nodes (conditions). */
export interface BoolExpr {
  readonly _tag: string;
}

// ── Leaf nodes ──────────────────────────────────────────────────────

/** Reference to a relation column. */
export interface Column extends Expr {
  readonly _tag: 'Column';
  readonly relation: string;
  readonly name: string;
  readonly refAlias?: string;
}

export function column(relation: string, name: string, refAlias?: string): Column {
  return { _tag: 'Column', relation, name, refAlias };
}

/** A constant value. */
export interface Literal extends Expr {
  readonly _tag: 'Literal';
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  readonly value: any;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function literal(value: any): Literal {
  return { _tag: 'Literal', value };
}

// ── Arithmetic ──────────────────────────────────────────────────────

/** Binary arithmetic: +, -, *, /, %. */
export interface Arithmetic extends Expr {
  readonly _tag: 'Arithmetic';
  readonly op: '+' | '-' | '*' | '/' | '%';
  readonly left: Expr;
  readonly right: Expr;
}

export function arithmetic(
  op: '+' | '-' | '*' | '/' | '%',
  left: Expr,
  right: Expr,
): Arithmetic {
  return { _tag: 'Arithmetic', op, left, right };
}

// ── Function call ───────────────────────────────────────────────────

/** Built-in function call: distance(V1, V2), upper(S), etc. */
export interface FuncCall extends Expr {
  readonly _tag: 'FuncCall';
  readonly name: string;
  readonly args: readonly Expr[];
}

export function funcCall(name: string, args: readonly Expr[]): FuncCall {
  return { _tag: 'FuncCall', name, args };
}

// ── Aggregation ─────────────────────────────────────────────────────

/** Aggregation expression: count<X>, sum<X>, top_k<k, ...>, etc. */
export interface AggExpr extends Expr {
  readonly _tag: 'AggExpr';
  readonly func: string;
  readonly column?: Expr;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  readonly params: readonly any[];
  readonly passthrough: readonly Expr[];
  readonly orderColumn?: Expr;
  readonly desc: boolean;
}

export function aggExpr(opts: {
  func: string;
  column?: Expr;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  params?: readonly any[];
  passthrough?: readonly Expr[];
  orderColumn?: Expr;
  desc?: boolean;
}): AggExpr {
  return {
    _tag: 'AggExpr',
    func: opts.func,
    column: opts.column,
    params: opts.params ?? [],
    passthrough: opts.passthrough ?? [],
    orderColumn: opts.orderColumn,
    desc: opts.desc ?? true,
  };
}

// ── Ordering ────────────────────────────────────────────────────────

/** A column with sort direction. */
export interface OrderedColumn extends Expr {
  readonly _tag: 'OrderedColumn';
  readonly column: Expr;
  readonly descending: boolean;
}

export function orderedColumn(col: Expr, descending: boolean): OrderedColumn {
  return { _tag: 'OrderedColumn', column: col, descending };
}

// ── Boolean expressions ─────────────────────────────────────────────

/** Binary comparison: ==, !=, <, <=, >, >=. */
export interface Comparison extends BoolExpr {
  readonly _tag: 'Comparison';
  readonly op: '=' | '!=' | '<' | '<=' | '>' | '>=';
  readonly left: Expr;
  readonly right: Expr;
}

export function comparison(
  op: '=' | '!=' | '<' | '<=' | '>' | '>=',
  left: Expr,
  right: Expr,
): Comparison {
  return { _tag: 'Comparison', op, left, right };
}

/** Logical AND of two conditions (IQL comma). */
export interface And extends BoolExpr {
  readonly _tag: 'And';
  readonly left: BoolExpr;
  readonly right: BoolExpr;
}

export function and(left: BoolExpr, right: BoolExpr): And {
  return { _tag: 'And', left, right };
}

/** Logical OR - requires splitting into multiple queries. */
export interface Or extends BoolExpr {
  readonly _tag: 'Or';
  readonly left: BoolExpr;
  readonly right: BoolExpr;
}

export function or(left: BoolExpr, right: BoolExpr): Or {
  return { _tag: 'Or', left, right };
}

/** Negation: !relation(X, Y) in IQL. */
export interface Not extends BoolExpr {
  readonly _tag: 'Not';
  readonly operand: BoolExpr;
}

export function not(operand: BoolExpr): Not {
  return { _tag: 'Not', operand };
}

/** Membership test: Column appears in another relation's column. */
export interface InExpr extends BoolExpr {
  readonly _tag: 'InExpr';
  readonly column: Expr;
  readonly targetColumn: Expr;
}

export function inExpr(col: Expr, target: Expr): InExpr {
  return { _tag: 'InExpr', column: col, targetColumn: target };
}

/** Negated membership test. */
export interface NegatedIn extends BoolExpr {
  readonly _tag: 'NegatedIn';
  readonly column: Expr;
  readonly targetColumn: Expr;
}

export function negatedIn(col: Expr, target: Expr): NegatedIn {
  return { _tag: 'NegatedIn', column: col, targetColumn: target };
}

/** Multi-column negation/existence check against a relation. */
export interface MatchExpr extends BoolExpr {
  readonly _tag: 'MatchExpr';
  readonly relation: string;
  readonly bindings: Record<string, Expr>;
  readonly negated: boolean;
}

export function matchExpr(
  relation: string,
  bindings: Record<string, Expr>,
  negated: boolean,
): MatchExpr {
  return { _tag: 'MatchExpr', relation, bindings, negated };
}

// ── Type guards ─────────────────────────────────────────────────────

export function isColumn(e: Expr): e is Column {
  return e._tag === 'Column';
}
export function isLiteral(e: Expr): e is Literal {
  return e._tag === 'Literal';
}
export function isArithmetic(e: Expr): e is Arithmetic {
  return e._tag === 'Arithmetic';
}
export function isFuncCall(e: Expr): e is FuncCall {
  return e._tag === 'FuncCall';
}
export function isAggExpr(e: Expr): e is AggExpr {
  return e._tag === 'AggExpr';
}
export function isOrderedColumn(e: Expr): e is OrderedColumn {
  return e._tag === 'OrderedColumn';
}

export function isComparison(e: BoolExpr): e is Comparison {
  return e._tag === 'Comparison';
}
export function isAnd(e: BoolExpr): e is And {
  return e._tag === 'And';
}
export function isOr(e: BoolExpr): e is Or {
  return e._tag === 'Or';
}
export function isNot(e: BoolExpr): e is Not {
  return e._tag === 'Not';
}
export function isInExpr(e: BoolExpr): e is InExpr {
  return e._tag === 'InExpr';
}
export function isNegatedIn(e: BoolExpr): e is NegatedIn {
  return e._tag === 'NegatedIn';
}
export function isMatchExpr(e: BoolExpr): e is MatchExpr {
  return e._tag === 'MatchExpr';
}
